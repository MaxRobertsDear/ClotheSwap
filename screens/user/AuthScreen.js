import React, { useState, useEffect, useReducer, useCallback } from 'react'
import {
  Button,
  View,
  KeyboardAvoidingView,
  StyleSheet,
  ActivityIndicator,
  Alert,
  Platform,
} from 'react-native'
import { LinearGradient } from 'expo-linear-gradient'
import { useDispatch } from 'react-redux'

import Colors from '../../constants/Colors'
import Input from '../../components/UI/Input'
import Card from '../../components/UI/Card'
import * as authActions from '../../store/actions/auth'

const FORM_INPUT_UPDATE = 'FORM_INPUT_UPDATE'

const formReducer = (state, action) => {
  if (action.type === FORM_INPUT_UPDATE) {
    const updatedValues = {
      ...state.inputValues,
      [action.input]: action.value,
    }
    const updatedValidities = {
      ...state.inputValidities,
      [action.input]: action.isValid,
    }
    let updatedFormIsValid = true
    for (const key in updatedValidities) {
      updatedFormIsValid = updatedFormIsValid && updatedValidities[key]
    }
    return {
      formIsValid: updatedFormIsValid,
      inputValidities: updatedValidities,
      inputValues: updatedValues,
    }
  }
  return state
}

const AuthScreen = () => {
  const dispatch = useDispatch()
  const [isSignup, setIsSignup] = useState(false)
  const [isLoading, setIsLoading] = useState(false)
  const [error, setError] = useState()

  const [formState, dispatchFormState] = useReducer(formReducer, {
    inputValues: {
      email: '',
      password: '',
    },
    inputValidities: {
      email: true,
      password: true,
    },
    formIsValid: false,
  })

  useEffect(() => {
    if (error) {
      Alert.alert('An error occurred', error, [{ text: 'Okay' }])
    }
  }, [error])

  const signupLoginHandler = async () => {
    if (formState.formIsValid) {
      let action
      if (isSignup) {
        action = authActions.signup(
          formState.inputValues.email,
          formState.inputValues.password,
        )
      } else if (!isSignup) {
        action = authActions.login(
          formState.inputValues.email,
          formState.inputValues.password,
        )
      }
      setIsLoading(true)
      try {
        await dispatch(action)
      } catch (err) {
        setError(err.message)
        setIsLoading(false)
      }
    }
    return
  }

  const textChangeHandler = useCallback(
    (inputIdentifier, text) => {
      const emailRegex = /^(([^<>()\[\]\\.,;:\s@"]+(\.[^<>()\[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/
      let isValid = false
      if (inputIdentifier === 'password' && text.trim().length > 5) {
        isValid = true
      }
      if (inputIdentifier === 'email' && emailRegex.test(text.toLowerCase())) {
        isValid = true
      }
      dispatchFormState({
        type: FORM_INPUT_UPDATE,
        value: text,
        isValid: isValid,
        input: inputIdentifier,
      })
    },
    [dispatchFormState],
  )

  return (
    <KeyboardAvoidingView
      behavior={Platform.OS === 'ios' ? 'padding' : 'height'}
      style={styles.screen}
    >
      <LinearGradient colors={['#ffafbd', '#ffc3a0']} style={styles.gradient}>
        <Card style={styles.authContainer}>
          <Input
            label='Email'
            placeholder='example@example.com'
            keyboardType='email-address'
            autoCapitalize='none'
            returnKeyType='next'
            errorText={formState.inputValidities.email}
            value={formState.inputValues.email}
            onChangeText={(input) => textChangeHandler('email', input)}
          />
          <Input
            label='Password'
            placeholder='minimum 6 characters'
            keyboardType='default'
            autoCapitalize='none'
            secureTextEntry
            minLength={5}
            errorText={formState.inputValidities.password}
            value={formState.inputValues.password}
            onChangeText={(input) => textChangeHandler('password', input)}
          />
          <View style={styles.buttonContainer}>
            {isLoading ? (
              <ActivityIndicator size='small' color={Colors.primary} />
            ) : (
              <Button
                title={isSignup ? 'Sign Up' : 'Login'}
                color={Colors.primary}
                onPress={signupLoginHandler}
              />
            )}
            <Button
              title={`Switch to ${isSignup ? 'Login' : 'Sign Up'}`}
              color={Colors.accent}
              onPress={() => {
                setIsSignup((prevState) => !prevState)
              }}
            />
          </View>
        </Card>
      </LinearGradient>
    </KeyboardAvoidingView>
  )
}

const styles = StyleSheet.create({
  screen: {
    flex: 1,
  },
  gradient: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
  },
  authContainer: {
    padding: 20,
    width: '80%',
    maxWidth: 400,
    maxHeight: 400,
  },
  buttonContainer: {
    marginTop: 10,
  },
})

export default AuthScreen
