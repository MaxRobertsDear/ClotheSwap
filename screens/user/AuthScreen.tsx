import React, { useState, useEffect, useReducer, useCallback } from 'react'
import {
  Button,
  View,
  KeyboardAvoidingView,
  StyleSheet,
  ActivityIndicator,
  Alert,
  Platform,
} from 'react-native'
import { LinearGradient } from 'expo-linear-gradient'
import { useDispatch } from 'react-redux'

import Colors from '../../constants/Colors'
import Input from '../../components/UI/Input'
import Card from '../../components/UI/Card'
import * as authActions from '../../store/actions/auth'

const FORM_INPUT_UPDATE = 'FORM_INPUT_UPDATE'

// interface iInputValues {
//   email: string;
//   password: string;
// }

interface iInputValidities {
  [key: string]: boolean;
}

const formReducer = (
  state: {
    // inputValues: iInputValues,
    inputValidities: iInputValidities,
    formIsValid: boolean,
  },
  action: { type: string, input: string, isValid: boolean },
) => {
  if (action.type === FORM_INPUT_UPDATE) {
    // const updatedValues = {
    //   ...state.inputValues,
    //   [action.input]: action.value,
    // }
    const updatedValidities = {
      ...state.inputValidities,
      [action.input]: action.isValid,
    }
    let updatedFormIsValid = true
    for (const key in updatedValidities) {
      updatedFormIsValid = updatedFormIsValid && updatedValidities[key]
    }
    return {
      formIsValid: updatedFormIsValid,
      inputValidities: updatedValidities,
      // inputValues: updatedValues,
    }
  }
  return state
}

const AuthScreen = () => {
  const dispatch = useDispatch()
  const [isSignup, setIsSignup] = useState(false)
  const [isLoading, setIsLoading] = useState(false)
  const [error, setError] = useState()
  const [email, setEmail] = useState('')
  const [password, setPassword] = useState('')

  const [formState, dispatchFormState] = useReducer(formReducer, {
    // inputValues: {
    //   email: '',
    //   password: '',
    // },
    inputValidities: {
      email: true,
      password: true,
    },
    formIsValid: false,
  })

  useEffect(() => {
    if (error) {
      Alert.alert('An error occurred', error, [{ text: 'Okay' }])
    }
  }, [error])

  const signupLoginHandler = async () => {
    if (formState.formIsValid) {
      let action
      if (isSignup) {
        action = authActions.signup({
          email: email,
          password: password,
        })
      } else if (!isSignup) {
        action = authActions.login({
          email: email,
          password: password,
        })
      }
      setIsLoading(true)
      try {
        await dispatch(action)
      } catch (err) {
        setError(err.message)
        setIsLoading(false)
      }
    }
    return
  }

  const formFieldValidator = useCallback(
    (inputIdentifier, text) => {
      const emailRegex = /^(([^<>()\[\]\\.,;:\s@"]+(\.[^<>()\[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/
      let isValid = false
      if (inputIdentifier === 'password' && text.trim().length > 5) {
        isValid = true
      }
      if (inputIdentifier === 'email' && emailRegex.test(text.toLowerCase())) {
        isValid = true
      }
      dispatchFormState({
        type: FORM_INPUT_UPDATE,
        // value: text,
        isValid: isValid,
        input: inputIdentifier,
      })
    },
    [dispatchFormState],
  )

  return (
    <KeyboardAvoidingView
      behavior={Platform.OS === 'ios' ? 'padding' : 'height'}
      style={styles.screen}
    >
      <LinearGradient colors={['#ffafbd', '#ffc3a0']} style={styles.gradient}>
        <Card style={styles.authContainer}>
          <Input
            label='Email'
            placeholder='example@example.com'
            keyboardType='email-address'
            autoCapitalize='none'
            returnKeyType='next'
            errorText={formState.inputValidities.email}
            value={email}
            onChangeText={(input: string) => setEmail(input)}
            onBlur={() => formFieldValidator('email', email)}
          />
          <Input
            label='Password'
            placeholder='minimum 6 characters'
            keyboardType='default'
            autoCapitalize='none'
            secureTextEntry
            minLength={5}
            errorText={formState.inputValidities.password}
            value={password}
            onChangeText={(input: string) => setPassword(input)}
            onBlur={() => formFieldValidator('password', password)}
          />
          <View style={styles.buttonContainer}>
            {isLoading ? (
              <ActivityIndicator size='small' color={Colors.primary} />
            ) : (
              <Button
                title={isSignup ? 'Sign Up' : 'Login'}
                color={Colors.primary}
                onPress={signupLoginHandler}
              />
            )}
            <Button
              title={`Switch to ${isSignup ? 'Login' : 'Sign Up'}`}
              color={Colors.accent}
              onPress={() => {
                setIsSignup((prevState) => !prevState)
              }}
            />
          </View>
        </Card>
      </LinearGradient>
    </KeyboardAvoidingView>
  )
}

const styles = StyleSheet.create({
  screen: {
    flex: 1,
  },
  gradient: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
  },
  authContainer: {
    padding: 20,
    width: '80%',
    maxWidth: 400,
    maxHeight: 400,
    shadowColor: 'black',
    shadowOpacity: 0.27,
    shadowOffset: { width: 0, height: 2 },
    shadowRadius: 8,
  },
  buttonContainer: {
    marginTop: 10,
  },
})

export default AuthScreen
