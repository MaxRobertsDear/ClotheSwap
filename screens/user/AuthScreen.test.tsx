import React from 'react'
import { Alert } from 'react-native'
import { fireEvent, waitFor } from 'react-native-testing-library'
import '@testing-library/jest-native/extend-expect'
import { renderWithRedux } from '../../_testing/custom.render'
import AuthScreen from './AuthScreen'

test('renders correctly', async () => {
  const tree = renderWithRedux(<AuthScreen />)
  expect(tree).toMatchSnapshot()
})
test('email and password fields correctly update with appropriate validation messages', async () => {
  const {
    queryByPlaceholder,
    queryByText,
    getByDisplayValue,
    getByText,
  } = renderWithRedux(<AuthScreen />)
  const passwordField = queryByPlaceholder('minimum 6 characters')
  const emailField = queryByPlaceholder('example@example.com')
  const emailValidationMessage = queryByText(/Please enter a valid Email/i)
  const passwordValidationMessage = queryByText(
    /Please enter a valid Password/i,
  )
  // no validation messages on screen start
  expect(emailValidationMessage).not.toBeTruthy()
  expect(passwordValidationMessage).not.toBeTruthy()

  // user fills out required fields correctly
  fireEvent.changeText(passwordField, '123456')
  fireEvent.changeText(emailField, 'jondoe@email.com')

  expect(getByDisplayValue('123456')).toBeTruthy()
  expect(getByDisplayValue('jondoe@email.com')).toBeTruthy()
  expect(emailValidationMessage).not.toBeTruthy()
  expect(passwordValidationMessage).not.toBeTruthy()

  // validation prompts shown on invalid entries
  fireEvent.changeText(passwordField, '123')
  fireEvent.changeText(emailField, 'email@email.com')
  fireEvent.changeText(emailField, '')

  expect(getByDisplayValue('123')).toBeTruthy()
  expect(getByDisplayValue('')).toBeTruthy()
  await waitFor(() => getByText(/Please enter a valid Password/i))
  await waitFor(() => getByText(/Please enter a valid Email/i))
})

test('user can toggle between login and signup', async () => {
  const { getAllByA11yRole, getByText } = renderWithRedux(<AuthScreen />)

  // user is presented with Login screen by default
  expect(getAllByA11yRole('button')[0]).toHaveTextContent(/Login/i)
  expect(getAllByA11yRole('button')[1]).toHaveTextContent(/Switch to Sign Up/i)

  // when toggling Login/Sign Up, user is presented with appropriate options
  fireEvent.press(getByText(/Switch to Sign Up/i))
  expect(getByText(/Sign Up/i)).toBeTruthy()
  expect(getByText(/Switch to Login/i)).toBeTruthy()
})

test('error message shown for invalid email / password', async () => {
  jest.spyOn(Alert, 'alert')
  const { queryByPlaceholder } = renderWithRedux(<AuthScreen />)
  const passwordField = queryByPlaceholder('minimum 6 characters')
  const emailField = queryByPlaceholder('example@example.com')

  // user tries to login with invalid email
  fireEvent.changeText(emailField, 'test@test.com')
  fireEvent.changeText(passwordField, '123456')
  // TODO: mock the api response for success and failure

  // fireEvent.press(getByText(/Login/i))
  // expect(Alert.alert).toHaveBeenCalled()
})
