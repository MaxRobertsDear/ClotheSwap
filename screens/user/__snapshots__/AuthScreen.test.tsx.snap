// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  onLayout={[Function]}
  style={
    Array [
      Object {
        "flex": 1,
      },
      Object {
        "paddingBottom": 0,
      },
    ]
  }
>
  <ViewManagerAdapter_ExpoLinearGradient
    proxiedProperties={
      Object {
        "colors": Array [
          4294946749,
          4294951840,
        ],
        "endPoint": undefined,
        "locations": undefined,
        "startPoint": undefined,
      }
    }
    style={
      Object {
        "alignItems": "center",
        "flex": 1,
        "justifyContent": "center",
      }
    }
  >
    <View
      style={
        Object {
          "backgroundColor": "white",
          "borderRadius": 10,
          "elevation": 5,
          "maxHeight": 400,
          "maxWidth": 400,
          "padding": 20,
          "shadowColor": "black",
          "shadowOffset": Object {
            "height": 2,
            "width": 0,
          },
          "shadowOpacity": 0.27,
          "shadowRadius": 8,
          "width": "80%",
        }
      }
    >
      <View
        style={
          Object {
            "width": "100%",
          }
        }
      >
        <Text
          style={
            Object {
              "fontFamily": "open-sans-bold",
              "marginVertical": 8,
            }
          }
        >
          Email
        </Text>
        <TextInput
          allowFontScaling={true}
          autoCapitalize="none"
          keyboardType="email-address"
          onBlur={[Function]}
          onChangeText={[Function]}
          placeholder="example@example.com"
          rejectResponderTermination={true}
          returnKeyType="next"
          style={
            Object {
              "borderBottomColor": "#ccc",
              "borderBottomWidth": 1,
              "paddingHorizontal": 2,
              "paddingVertical": 5,
            }
          }
          underlineColorAndroid="transparent"
          value=""
        />
      </View>
      <View
        style={
          Object {
            "width": "100%",
          }
        }
      >
        <Text
          style={
            Object {
              "fontFamily": "open-sans-bold",
              "marginVertical": 8,
            }
          }
        >
          Password
        </Text>
        <TextInput
          allowFontScaling={true}
          autoCapitalize="none"
          keyboardType="default"
          minLength={5}
          onBlur={[Function]}
          onChangeText={[Function]}
          placeholder="minimum 6 characters"
          rejectResponderTermination={true}
          secureTextEntry={true}
          style={
            Object {
              "borderBottomColor": "#ccc",
              "borderBottomWidth": 1,
              "paddingHorizontal": 2,
              "paddingVertical": 5,
            }
          }
          underlineColorAndroid="transparent"
          value=""
        />
      </View>
      <View
        style={
          Object {
            "marginTop": 10,
          }
        }
      >
        <View
          accessibilityRole="button"
          accessibilityState={Object {}}
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
        >
          <View
            style={
              Array [
                Object {},
              ]
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "#007AFF",
                    "fontSize": 18,
                    "margin": 8,
                    "textAlign": "center",
                  },
                  Object {
                    "color": "black",
                  },
                ]
              }
            >
              Login
            </Text>
          </View>
        </View>
        <View
          accessibilityRole="button"
          accessibilityState={Object {}}
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
        >
          <View
            style={
              Array [
                Object {},
              ]
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "#007AFF",
                    "fontSize": 18,
                    "margin": 8,
                    "textAlign": "center",
                  },
                  Object {
                    "color": "#FFC107",
                  },
                ]
              }
            >
              Switch to Sign Up
            </Text>
          </View>
        </View>
      </View>
    </View>
  </ViewManagerAdapter_ExpoLinearGradient>
</View>
`;
